### PR TITLE
[python-package] early stopping min_delta (fixes #2526)

### LINKED ISSUE #2526: Feature Request: customizable early_stopping_tolerance
I have a situation where the default numerical tolerance (0.001) for early stopping is too large. My target has a gamma distribution and the LGB Regressor reaches convergence too early, when the numerous low target values are well approximated by the model, but the few large values are still underestimated. When I deactivate early stopping, I can see the loss metric still improving at the 4th or more decimal digit, past the best iteration reached during early stopping. It would be great to be able to set manually the tolerance.

### PR DESCRIPTION
**Pull request by jmoralez**
This aims to close #2526 by providing an additional argument `min_delta` to the early stopping callback, which allows to perform early stopping when the score doesn't improve by at least that `min_delta` in `stopping_rounds` iterations. The default value is 0, so the current behavior is maintained.

In the case for multiple metrics the user can provide a single delta for all metrics or a specific delta for each metric. The handling of all the cases (single metric, multiple metrics, multiple metrics considering first only) can be a bit cumbersome so I'm open to suggestions on how to handle them.

### CODE REVIEW

**Review by jameslamb — COMMENTED** [review]
Thanks for working on this! I left a few preliminary comments.

**Comment by jameslamb** [comment]
@jmoralez by the way, since you just joined...we use a bot to create release notes automatically based on the set of pull requests merged between git tags.

That bot chooses which section of the release notes to put a change in based on the use of specific labels.

image

https://github.com/microsoft/LightGBM/blob/3942126592120d2372c5237cae67064ac5d2000e/.github/release-drafter.yml#L4-L15

So every PR needs to be assigned a label from that set. I'd call this one `feature`. If you're ever unsure, `maintenance` is a catch-all.

File: python-package/lightgbm/engine.py [hunk 1]
```diff
@@ -421,8 +424,8 @@ def cv(params, train_set, num_boost_round=100,
        folds=None, nfold=5, stratified=True, shuffle=True,
        metrics=None, fobj=None, feval=None, init_model=None,
        feature_name='auto', categorical_feature='auto',
-       early_stopping_rounds=None, fpreproc=None,
-       verbose_eval=None, show_stdv=True, seed=0,
+       early_stopping_rounds=None, early_stopping_threshold=0.0,
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply python-package/lightgbm/engine.py:427]: I guess this PR was developed before #4574 where it is proposed to eliminate advanced arguments of high-level functions and encourage users to use callbacks directly. If that proposal will be accepted, I think `early_stopping_threshold` shouldn't be exposed in training API.

[the remark this answers is not in the document]
Comment (jmoralez) [reply python-package/lightgbm/engine.py:427]: Yes, I started working on this earlier but I agree with #4574 and I think it will also be easier to test (just testing the callback on one interface instead of testing the argument in `train`, `cv` and the scikit-learn estimators) and to document. I'll update this PR accordingly.

**Comment by jmoralez** [comment]
I can give it a shot once this is done.

File: python-package/lightgbm/engine.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply python-package/lightgbm/engine.py:427]: oh good point, I didn't make that connection about #4574. Thanks @StrikerRUS

**Comment by StrikerRUS** [comment]
In general the proposed solution looks good to me. Thanks for the PR!

In https://github.com/dmlc/xgboost/pull/7137#issuecomment-888949278 it is said that `min_delta` is a commonly used name for a such type of parameter. Maybe it is better to use name `min_delta` here as well for the consistency with other ML libraries?
> Also, looking into APIs of early stopping callbacks from other frameworks (for example, [pytorch-lightning](https://pytorch-lightning.readthedocs.io/en/latest/api/pytorch_lightning.callbacks.early_stopping.html#pytorch_lightning.callbacks.early_stopping.EarlyStopping) and [tensorflow](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks/EarlyStopping#args)), I see only one parameter `min_delta` and no tolerance parameter.

I believe all `_log_warning()` occurrences should be replaced with `_log_info()` because those messages look like debug-level info, not like something that is expected to be fixed at the user side.

Please enhance new parameter description with some info about the semantics of expected parameter types. It should help users avoid this error at the time of writing their code but not after they read this error message.
```
Must provide a single early stopping threshold or as many as metrics.
```
Something like the following:
```
    If float, this singe value is used as early stopping threshold for all metrics.
    If list, its length should match the total number of metrics to use one early stopping threshold per one metric.
```

**Comment by jmoralez** [comment]
@StrikerRUS I've changed the name to `min_delta` and updated the description. Let me know what you think.

**Review by shiyu1994 — APPROVED** [review]
The changes LGTM. Thank you! @jmoralez

**Comment by jmoralez** [comment]
@StrikerRUS should I merge this?

**Comment by StrikerRUS** [comment]
@jmoralez 
> should I merge this?

TBH, I'm not sure. We decided to give some time (about 2 weeks or so) after the `v3.3.1` release for reporting critical bugs in it and during that time not merge large PRs that contain new features.

> I'd like to ask to **not** merge new large PRs with major features for one or two weeks after releasing `v3.3.0`. This will ease the process of creating `v3.3.1` hotfix release if that will be needed in case of some critical bugs in `v3.3.0` we might get reports about.
   https://github.com/microsoft/LightGBM/pull/4633#issuecomment-937778939

> I'd also like to propose that after the v3.3.1 release, we again wait 2 weeks to begin merging other large PRs, in case a v3.3.2 release is required.
   https://github.com/microsoft/LightGBM/pull/4715#issue-1035256945

This PR touches existing code and may possibly change current logic. I'd better wait a little bit before merging it.
@jameslamb @shiyu1994 WDYT?

**Comment by jameslamb** [comment]
> I'd better wait a little bit before merging it

I agree. v3.3.1 was only merged 6 days ago (https://github.com/microsoft/LightGBM/pull/4715#event-5524801866), so I think we should wait another week for this PR.

**Comment by shiyu1994** [comment]
>I'd better wait a little bit before merging it.

I agree.

**Comment by jameslamb** [comment]
Great work on this, @jmoralez !!!
